src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -968,21 +968,40 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
                             _rcwCache.Remove(identity);
                         }
                     }
+                }
+
+                if (wrapperMaybe is not null)
+                {
+                    retValue = wrapperMaybe;
 
-                    if (wrapperMaybe is not null)
+                    using (_lock.EnterScope())
                     {
-                        retValue = wrapperMaybe;
-                        NativeObjectWrapper wrapper = NativeObjectWrapper.Create(
-                            identity,
-                            inner,
-                            this,
-                            retValue,
-                            flags);
-                        if (!s_rcwTable.TryAdd(retValue, wrapper))
+                        if (!s_rcwTable.TryGetValue(retValue, out NativeObjectWrapper? wrapper))
                         {
-                            wrapper.Release();
-                            throw new NotSupportedException();
+                            // Exit the lock before we create the native object wrapper
+                            // as we may execute arbitrary user code.
+                            _lock.Exit();
+                            try
+                            {
+                                NativeObjectWrapper newWrapper = NativeObjectWrapper.Create(
+                                    identity,
+                                    inner,
+                                    this,
+                                    retValue,
+                                    flags);
+
+                                wrapper = s_rcwTable.GetValue(retValue, _ => newWrapper);
+                                if (wrapper != newWrapper)
+                                {
+                                    newWrapper.Release();
+                                }
+                            }
+                            finally
+                            {
+                                _lock.Enter();
+                            }
                         }
+
                         _rcwCache.Add(identity, wrapper._proxyHandle);
```

Comment:
Can you explain why we need to enter the lock again before adding the wrapper to the RCW cache? What I'm not following is: if it's important that we do operations on the RCW table and the RCW cache atomically, then why are we doing `s_rcwTable.GetValue` with the new wrapper **before** entering the lock again? Basically, why don't we enter the lock again right after creating the new wrapper? If instead it's not needed to mutate the RCW table and the RCW cache atomically, then what value does taking the lock specifically just for this one call on the RCW cache provide? What problem is it solving? 🤔

(I guess it also wouldn't hurt to add comments that answer these questions, for future reference)

Reply:
The RCW cache isn't thread-safe (it's a regular `Dictionary`), so we need to synchronize around editing it. I'll add some comments.